### PR TITLE
modify the column type change doc

### PR DESCRIPTION
**Pull request by AilinKid**
Signed-off-by: AilinKid <314806019@qq.com>

### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

modify some lines about the column type change

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by qiancai** [comment]
@AilinKid would you please add technical reviewers for this PR? Thanks!

File: sql-statements/sql-statement-modify-column.md [hunk 1]
```diff
@@ -8,6 +8,8 @@ aliases: ['/docs-cn/dev/sql-statements/sql-statement-modify-column/','/docs-cn/d
 
 `ALTER TABLE .. MODIFY COLUMN` 语句用于修改已有表上的列，包括列的数据类型和属性。若要同时重命名，可改用 [`CHANGE COLUMN`](/sql-statements/sql-statement-change-column.md) 语句。
 
+在 v5.0.2 及其之后的版本中，类型变更将支持 Reorg 类型变更，包括但是不仅限于 varchar -> bigint、decimal 精度修改、varchar(10) -> varchar(5) 长度压缩等类型的变更。
```
Comment (zimulala) [inline sql-statements/sql-statement-modify-column.md:11]: 应该是 5.1.0 吧。

**Comment by AilinKid** [comment]
Do not merge temporarily for now.

File: sql-statements/sql-statement-modify-column.md [hunk 2]
```diff
@@ -104,21 +172,29 @@ Create Table: CREATE TABLE `t1` (
...
+    ERROR 8200 (HY000): Unsupported modify column: column is generated
     ```
 
-* 不支持修改 decimal 类型的精度。例如：
-
-    ```sql
-    CREATE TABLE t (a DECIMAL(5, 3));
-    ALTER TABLE t MODIFY COLUMN a DECIMAL(6, 3);
-    ERROR 8200 (HY000): Unsupported modify column: can't change decimal column precision
+* 不支持修一部分类型 (部分时间类型、Bit、Json 等) 上的变更。例如：
```
Comment (zimulala) [inline sql-statements/sql-statement-modify-column.md:192]: 要添上 enum 和 set 类型？可以说明一下原因

File: sql-statements/sql-statement-modify-column.md [hunk 3]
```diff
@@ -104,21 +172,29 @@ Create Table: CREATE TABLE `t1` (
...
+* 不支持修改主键列上的类型。例如：
+  
+    ```
+    CREATE TABLE t (a int primary key);
+    ALTER TABLE t MODIFY COLUMN a VARCHAR(10);
+    ERROR 8200 (HY000): Unsupported modify column: column has primary key flag
+    ```
 
-* 不支持有损变更，以及部分数据类型的更改（包括整数改为字符串或 BLOB 格式等）。例如：
+* 不支持修改 generated column 的类型。例如：
```
Comment (zimulala) [inline sql-statements/sql-statement-modify-column.md:184]: 加一下 partition？

**Review by zimulala — COMMENTED** [review]
与 MySQL 兼容性问题那个文档也修改一下？

**Comment by zimulala** [comment]
还有个 sql-statement-change-column.go 文档需要更新

File: mysql-compatibility.md [hunk 4]
```diff
@@ -96,13 +96,10 @@ TiDB 主要使用 Prometheus 和 Grafana 来存储及查询相关的性能监控
 TiDB 中，所有支持的 DDL 变更操作都是在线执行的。与 MySQL 相比，TiDB 中的 DDL 存在以下限制：
 
 * 不能在单条 `ALTER TABLE` 语句中完成多个操作。例如，不能在单个语句中添加多个列或索引，否则，可能会输出 `Unsupported multi schema change` 的错误。
-* 不支持不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
-* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
-* 不支持将字段类型修改为其超集，例如不支持从 `INTEGER` 修改为 `VARCHAR`，或者从 `TIMESTAMP` 修改为 `DATETIME`，否则可能输出的错误信息 `Unsupported modify column: type %d not match origin %d`。
-* 更改/修改数据类型时，尚未支持“有损更改”，例如不支持从 BIGINT 更改为 INT。
-* 更改/修改 DECIMAL 类型时，不支持更改精度。
-* 更改/修改整数列时，不允许更改 `UNSIGNED` 属性。
+* `ALTER TABLE` 不支持一些类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
```
Comment (zimulala) [inline mysql-compatibility.md:99]:
```suggestion
* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
```

File: sql-statements/sql-statement-change-column.md [hunk 5]
```diff
@@ -93,39 +95,64 @@ Query OK, 0 rows affected (0.08 sec)
...
 ## MySQL 兼容性
 
 * 不支持在单个 `ALTER TABLE` 语句中进行多个更改。
-* 不支持有损变更，比如从 `BIGINT` 变为 INTEGER，或者从 `VARCHAR(255)` 变为 `VARCHAR(10)`。
-* 不支持修改 `DECIMAL` 类型的精度。
-* 不支持更改 `UNSIGNED` 属性。
+* 不支持主键列上的类型变更。
+* 不支持分区表上的列类型变更。
+* 不支持 generated column 上的列类型变更。
+* 不支持修一部分类型 (部分时间类型、Bit、Set、Enum、Json 等) 上的变更，缘由是 TiDB cast 函数和 MySQL 的行为有一些兼容性问题。
```
Comment (zimulala) [inline sql-statements/sql-statement-change-column.md:155]: s/Json/JSON

Comment (AilinKid) [inline sql-statements/sql-statement-change-column.md:155]:
```suggestion
* 不支持修一部分类型 (部分时间类型、Bit、Set、Enum、JSON 等) 上的变更，缘由是 TiDB cast 函数和 MySQL 的行为有一些兼容性问题。
```

File: sql-statements/sql-statement-modify-column.md [hunk 6]
```diff
@@ -104,21 +175,37 @@ Create Table: CREATE TABLE `t1` (
...
-* 不支持修改 decimal 类型的精度。例如：
+* 不支持修改分区表上的列类型。例如：
+
+    ```sql
+    CREATE TABLE t (c1 INT, c2 INT, c3 INT) partition by range columns(c1) ( partition p0 values less than (10), partition p1 values less than (maxvalue));
+    ALTER TABLE t MODIFY COLUMN c1 DATETIME;
+    ERROR 8200 (HY000): Unsupported modify column: table is partition table
+    ```
 
+* 不支持部分数据类型（例如，部分时间类型、Bit、Set、Enum、Json 等）的变更，因为 TiDB cast 函数与 MySQL 的行为有一些兼容性问题。例如：
```
Comment (AilinKid) [inline sql-statements/sql-statement-modify-column.md:203]:
```suggestion
* 不支持部分数据类型（例如，部分时间类型、Bit、Set、Enum、JSON 等）的变更，因为 TiDB cast 函数与 MySQL 的行为有一些兼容性问题。例如：
```

File: mysql-compatibility.md [hunk 7]
```diff
@@ -96,13 +96,10 @@ TiDB 主要使用 Prometheus 和 Grafana 来存储及查询相关的性能监控
 TiDB 中，所有支持的 DDL 变更操作都是在线执行的。与 MySQL 相比，TiDB 中的 DDL 存在以下限制：
 
 * 不能在单条 `ALTER TABLE` 语句中完成多个操作。例如，不能在单个语句中添加多个列或索引，否则，可能会输出 `Unsupported multi schema change` 的错误。
-* 不支持不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
-* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
-* 不支持将字段类型修改为其超集，例如不支持从 `INTEGER` 修改为 `VARCHAR`，或者从 `TIMESTAMP` 修改为 `DATETIME`，否则可能输出的错误信息 `Unsupported modify column: type %d not match origin %d`。
-* 更改/修改数据类型时，尚未支持“有损更改”，例如不支持从 BIGINT 更改为 INT。
-* 更改/修改 DECIMAL 类型时，不支持更改精度。
-* 更改/修改整数列时，不允许更改 `UNSIGNED` 属性。
+* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
```
Comment (zimulala) [inline mysql-compatibility.md:99]:
```suggestion
* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-change-column.md )。
```

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by zimulala** [comment]
PTAL @TomShawn

File: sql-statements/sql-statement-change-column.md [hunk 8]
```diff
@@ -93,39 +95,64 @@ Query OK, 0 rows affected (0.08 sec)
...
 ## MySQL 兼容性
 
 * 不支持在单个 `ALTER TABLE` 语句中进行多个更改。
-* 不支持有损变更，比如从 `BIGINT` 变为 INTEGER，或者从 `VARCHAR(255)` 变为 `VARCHAR(10)`。
-* 不支持修改 `DECIMAL` 类型的精度。
-* 不支持更改 `UNSIGNED` 属性。
+* 不支持主键列上的类型变更。
+* 不支持分区表上的列类型变更。
+* 不支持 generated column 上的列类型变更。
+* 不支持修一部分类型 (部分时间类型、Bit、Set、Enum、JSON 等) 上的变更，缘由是 TiDB cast 函数和 MySQL 的行为有一些兼容性问题。
```
Comment (qiancai) [inline sql-statements/sql-statement-change-column.md:155]:
```suggestion
* 不支持部分数据类型（例如，部分时间类型、Bit、Set、Enum、JSON 等）的变更，因为 TiDB cast 函数与 MySQL 的行为有一些兼容性问题。
```

File: sql-statements/sql-statement-modify-column.md [hunk 9]
```diff
@@ -96,6 +106,71 @@ Create Table: CREATE TABLE `t1` (
...
+
+> - 由于和 Async Commit 功能兼容，DDL 在开始进入到 Reorg Data 前会有一定时间 (约 2.5s) 的等待处理：
+
+>   ```
+>   Query OK, 0 rows affected (2.52 sec)
+>   ```
+
 ## MySQL 兼容性
 
 * 不支持在单个 `ALTER TABLE` 语句修改多个列，例如：
```
Comment (qiancai) [inline sql-statements/sql-statement-modify-column.md:176]:
```suggestion
* 不支持使用单个 `ALTER TABLE` 语句修改多个列，例如：
```

File: mysql-compatibility.md [hunk 10]
```diff
@@ -96,13 +96,10 @@ TiDB 主要使用 Prometheus 和 Grafana 来存储及查询相关的性能监控
...
 * 不能在单条 `ALTER TABLE` 语句中完成多个操作。例如，不能在单个语句中添加多个列或索引，否则，可能会输出 `Unsupported multi schema change` 的错误。
-* 不支持不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
-* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
-* 不支持将字段类型修改为其超集，例如不支持从 `INTEGER` 修改为 `VARCHAR`，或者从 `TIMESTAMP` 修改为 `DATETIME`，否则可能输出的错误信息 `Unsupported modify column: type %d not match origin %d`。
-* 更改/修改数据类型时，尚未支持“有损更改”，例如不支持从 BIGINT 更改为 INT。
-* 更改/修改 DECIMAL 类型时，不支持更改精度。
-* 更改/修改整数列时，不允许更改 `UNSIGNED` 属性。
+* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-modify-column.md)。
 * TiDB 中，`ALGORITHM={INSTANT,INPLACE,COPY}` 语法只作为一种指定，并不更改 `ALTER` 算法，详情参阅 [`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
+* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
```
Comment (qiancai) [inline mysql-compatibility.md:101]:
```suggestion
* 不支持添加或删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
```

File: mysql-compatibility.md [hunk 11]
```diff
@@ -96,13 +96,10 @@ TiDB 主要使用 Prometheus 和 Grafana 来存储及查询相关的性能监控
 TiDB 中，所有支持的 DDL 变更操作都是在线执行的。与 MySQL 相比，TiDB 中的 DDL 存在以下限制：
 
 * 不能在单条 `ALTER TABLE` 语句中完成多个操作。例如，不能在单个语句中添加多个列或索引，否则，可能会输出 `Unsupported multi schema change` 的错误。
-* 不支持不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
-* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
-* 不支持将字段类型修改为其超集，例如不支持从 `INTEGER` 修改为 `VARCHAR`，或者从 `TIMESTAMP` 修改为 `DATETIME`，否则可能输出的错误信息 `Unsupported modify column: type %d not match origin %d`。
-* 更改/修改数据类型时，尚未支持“有损更改”，例如不支持从 BIGINT 更改为 INT。
-* 更改/修改 DECIMAL 类型时，不支持更改精度。
-* 更改/修改整数列时，不允许更改 `UNSIGNED` 属性。
+* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节可以参考[`ALTER TABLE`](/sql-statements/sql-statement-modify-column.md)。
```
Comment (qiancai) [inline mysql-compatibility.md:99]:
```suggestion
* `ALTER TABLE` 不支持少部分类型的变更。比如，TiDB 不支持从 `DECIMAL` 到 `DATE` 的变更。当遇到不支持的类型变更时，TiDB 将会报 `Unsupported modify column: type %d not match origin %d` 的错误。更多细节，请参考[`ALTER TABLE`](/sql-statements/sql-statement-modify-column.md)。
```

File: mysql-compatibility.md [hunk 12]
```diff
@@ -96,13 +96,10 @@ TiDB 主要使用 Prometheus 和 Grafana 来存储及查询相关的性能监控
...
 * TiDB 中，`ALGORITHM={INSTANT,INPLACE,COPY}` 语法只作为一种指定，并不更改 `ALTER` 算法，详情参阅 [`ALTER TABLE`](/sql-statements/sql-statement-alter-table.md)。
+* 不支持添加/删除 `CLUSTERED` 类型的主键。要了解关于 `CLUSTERED` 主键的详细信息，请参考[聚簇索引](/clustered-indexes.md)。
+* 不支持不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
```
Comment (qiancai) [inline mysql-compatibility.md:102]:
```suggestion
* 不支持指定不同类型的索引 (`HASH|BTREE|RTREE|FULLTEXT`)。若指定了不同类型的索引，TiDB 会解析并忽略这些索引。
```

File: sql-statements/sql-statement-modify-column.md [hunk 13]
```diff
@@ -96,6 +106,71 @@ Create Table: CREATE TABLE `t1` (
...
+  `col1` varchar(5) DEFAULT NULL,
+  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30001
+1 row in set (0.00 sec)
+```
+
+**注意:**
+
+> - 当所变更的类型与已经存在的数据行产生冲突时，TiDB 会进行报错处理。在上述例子中，TiDB 将进行如下报错：
+
```
Comment (qiancai) [inline sql-statements/sql-statement-modify-column.md]:
```suggestion
>
```

File: sql-statements/sql-statement-modify-column.md [hunk 14]
```diff
@@ -96,6 +106,71 @@ Create Table: CREATE TABLE `t1` (
...
+>   ```
+>   alter table t1 modify column col1 varchar(4);
+>   ERROR 1406 (22001): Data Too Long, field len 4, data len 5
+>   ```
+
```
Comment (qiancai) [inline sql-statements/sql-statement-modify-column.md]:
```suggestion
>
```

File: sql-statements/sql-statement-modify-column.md [hunk 15]
```diff
@@ -96,6 +106,71 @@ Create Table: CREATE TABLE `t1` (
...
+> - 由于和 Async Commit 功能兼容，DDL 在开始进入到 Reorg Data 前会有一定时间 (约 2.5s) 的等待处理：
+
```
Comment (qiancai) [inline sql-statements/sql-statement-modify-column.md]:
```suggestion
>
```

**Review by qiancai — APPROVED** [review]

**Pull request merged**